File: kani-cli/src/commands/audit_tokens.rs
use crate::error::CliError;
use regex::Regex;
use std::fs;
use std::path::{Path, PathBuf};

pub struct Violation {
    pub file: PathBuf,
    pub line: usize,
    pub literal: String,
}

struct Patterns {
    hex: Regex,
    rgb: Regex,
    tailwind: Regex,
}

impl Patterns {
    fn new() -> Self {
        let palette = [
            "gray", "zinc", "slate", "neutral", "stone", "red", "orange", "amber", "yellow",
            "lime", "green", "emerald", "teal", "cyan", "sky", "blue", "indigo", "violet",
            "purple", "fuchsia", "pink", "rose", "white", "black",
        ]
        .join("|");

        Self {
            hex: Regex::new(r"#(?:[0-9a-fA-F]{8}|[0-9a-fA-F]{6}|[0-9a-fA-F]{3})\b")
                .expect("valid hex regex"),
            rgb: Regex::new(r"rgb[a]?\s*\(|hsl[a]?\s*\(").expect("valid rgb regex"),
            tailwind: Regex::new(&format!(
                r"\b(?:text|bg|border|ring|fill|stroke|from|to|via|shadow|accent)-(?:{})(?:-\d+)?",
                palette
            ))
            .expect("valid tailwind regex"),
        }
    }

    fn matches(&self, line: &str) -> Vec<String> {
        let mut found = Vec::new();
        for m in self.hex.find_iter(line) {
            found.push(m.as_str().to_owned());
        }
        for m in self.rgb.find_iter(line) {
            found.push(m.as_str().trim_end_matches(['(', ' ']).to_owned());
        }
        for m in self.tailwind.find_iter(line) {
            found.push(m.as_str().to_owned());
        }
        found
    }
}

pub fn run(dir: &Path, check: bool, max: usize) -> Result<(), CliError> {
    let violations = scan(dir)?;
    for v in &violations {
        println!("{}:{}: {}", v.file.display(), v.line, v.literal);
    }
    let count = violations.len();
    if count == 0 {
        println!("No hard-coded colour literals found in {}.", dir.display());
        return Ok(());
    }
    println!("\n{count} violation(s) found (baseline max: {max}).");
    if check && count > max {
        return Err(CliError::Other(format!(
            "{count} hard-coded colour violation(s) exceeds baseline of {max} — \
             migrate new literals to design tokens (or lower the baseline as you clean up)"
        )));
    }
    Ok(())
}

pub fn scan(dir: &Path) -> Result<Vec<Violation>, CliError> {
    let patterns = Patterns::new();
    let mut violations = Vec::new();
    collect(dir, &patterns, &mut violations)?;
    Ok(violations)
}

fn collect(dir: &Path, patterns: &Patterns, out: &mut Vec<Violation>) -> Result<(), CliError> {
    let entries = fs::read_dir(dir)
        .map_err(|e| CliError::Other(format!("cannot read {}: {e}", dir.display())))?;
    let mut entries: Vec<_> = entries.flatten().collect();
    entries.sort_by_key(|e| e.file_name());

    for entry in entries {
        let path = entry.path();
        let name = entry.file_name();
        let skip = matches!(
            name.to_string_lossy().as_ref(),
            "dist" | "vendor" | "node_modules"
        );
        if skip {
            continue;
        }
        if path.is_dir() {
            collect(&path, patterns, out)?;
        } else if path.extension().is_some_and(|ext| ext == "js") {
            scan_js_file(&path, patterns, out)?;
        }
    }
    Ok(())
}

fn scan_js_file(
    path: &Path,
    patterns: &Patterns,
    out: &mut Vec<Violation>,
) -> Result<(), CliError> {
    let content = fs::read_to_string(path)
        .map_err(|e| CliError::Other(format!("cannot read {}: {e}", path.display())))?;

    // A file-level `audit-ignore-file` directive skips the whole file — for
    // modules that are definitionally colour sources (e.g. the theme palette).
    if content.contains("audit-ignore-file") {
        return Ok(());
    }

    for (idx, line) in content.lines().enumerate() {
        let line_num = idx + 1;
        // A line-level `audit-ignore` directive opts that line out — for
        // intentional literals (the pure-black reader, scrims, contrast text)
        // that have no semantic-token equivalent.
        if line.contains("audit-ignore") {
            continue;
        }
        for literal in patterns.matches(line) {
            out.push(Violation {
                file: path.to_owned(),
                line: line_num,
                literal,
            });
        }
    }
    Ok(())
}

File: scripts/audit-tokens.mjs
#!/usr/bin/env node

import { readdirSync, readFileSync, statSync } from 'node:fs';
import { join, extname } from 'node:path';

const argv = process.argv.slice(2);
const flag = (name, fallback) => {
  const i = argv.indexOf(name);
  return i === -1 ? fallback : argv[i + 1];
};
const DIR = flag('--dir', 'static/js');
const CHECK = argv.includes('--check');
const MAX = Number(flag('--max', '0'));

const PALETTE = [
  'gray', 'zinc', 'slate', 'neutral', 'stone', 'red', 'orange', 'amber', 'yellow',
  'lime', 'green', 'emerald', 'teal', 'cyan', 'sky', 'blue', 'indigo', 'violet',
  'purple', 'fuchsia', 'pink', 'rose', 'white', 'black',
].join('|');

// `\b` after the hex body matches Rust's regex crate: a 3, 6 or 8 digit run
// followed by a word boundary, so `#abcd` is not reported as `#abc`.
const HEX = /#(?:[0-9a-fA-F]{8}|[0-9a-fA-F]{6}|[0-9a-fA-F]{3})\b/g;
const FUNC = /rgb[a]?\s*\(|hsl[a]?\s*\(/g;
const TAILWIND = new RegExp(
  String.raw`\b(?:text|bg|border|ring|fill|stroke|from|to|via|shadow|accent)-(?:${PALETTE})(?:-\d+)?`,
  'g',
);

const SKIP_DIRS = new Set(['dist', 'vendor', 'node_modules']);

/** @param {string} line */
function literalsIn(line) {
  const found = [];
  for (const m of line.matchAll(HEX)) found.push(m[0]);
  // Reported without the trailing paren/space, matching the Rust output.
  for (const m of line.matchAll(FUNC)) found.push(m[0].replace(/[(\s]+$/, ''));
  for (const m of line.matchAll(TAILWIND)) found.push(m[0]);
  return found;
}

/** @param {string} dir @param {{file:string,line:number,literal:string}[]} out */
function walk(dir, out) {
  let entries;
  try {
    entries = readdirSync(dir).sort();
  } catch (e) {
    console.error(`cannot read ${dir}: ${e.message}`);
    process.exit(2);
  }
  for (const name of entries) {
    if (SKIP_DIRS.has(name)) continue;
    const path = join(dir, name);
    if (statSync(path).isDirectory()) walk(path, out);
    else if (extname(path) === '.js') scanFile(path, out);
  }
}

/** @param {string} path @param {{file:string,line:number,literal:string}[]} out */
function scanFile(path, out) {
  const content = readFileSync(path, 'utf8');
  if (content.includes('audit-ignore-file')) return;
  content.split('\n').forEach((line, i) => {
    if (line.includes('audit-ignore')) return;
    for (const literal of literalsIn(line)) {
      out.push({ file: path, line: i + 1, literal });
    }
  });
}

const violations = [];
walk(DIR, violations);
for (const v of violations) console.log(`${v.file}:${v.line}: ${v.literal}`);

if (violations.length === 0) {
  console.log(`No hard-coded colour literals found in ${DIR}.`);
  process.exit(0);
}

console.log(`\n${violations.length} violation(s) found (baseline max: ${MAX}).`);
if (CHECK && violations.length > MAX) {
  // `error:` prefix matches what cargo printed, so CI logs read the same.
  console.error(
    `error: ${violations.length} hard-coded colour violation(s) exceeds baseline of ${MAX} — ` +
    'migrate new literals to design tokens (or lower the baseline as you clean up)',
  );
  process.exit(1);
}
